lib/utilities/clean-remove.js

```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+var path = require('path');
+var fs = require('fs-extra');
+var Promise = require('../ext/promise');
+var walkUp = require('./walk-up-path');
+var stat = Promise.denodeify(fs.stat);
+var remove = Promise.denodeify(fs.remove);
+var readdir = Promise.denodeify(fs.readdir);
+
+function cleanRemove(fileInfo) {
+  // see if file exists to avoid wasting time
+  return stat(fileInfo.outputPath).then(function() {
+    return remove(fileInfo.outputPath).then(function() {
+      var paths = walkUp(fileInfo.displayPath).map(function(thePath) {
+        return path.join(fileInfo.outputBasePath, thePath);
+      });
+
+      var linkedPromise = paths.reduce(function(linkedPromise, thePath) {
+        return linkedPromise.then(function(wasShortCircuited) {
+          if (wasShortCircuited) {
+            // optimization that says since my child dir wasn't empty,
+            // I can't be empty, so keep skipping
+            return true;
+          }
+
+          // get list of files remaining in this dir
+          return readdir(thePath).then(function(paths) {
+            if (paths.length) {
+              // don't check parent dirs since this one isn't empty
+              return true;
+            }
+
+            return remove(thePath).then(function() {
+              return false;
+            });
+          });
+        });
+      }, Promise.resolve());
+
+      return linkedPromise;
+    });
+  }).catch(function() {
+    // you tried to destroy a blueprint without first generating it
+    // instead of trying to read dirs that don't exist
+    // swallow error and carry on
```

Comment:
It's not a a user error, just a `fs.stat` error in this case. This is a supported thing in the CLI, if you destroy something that doesn't exist, the UI pretends to delete it:

```
$ ember d template jfsd/sdfljks/sdfsjl
uninstalling template
  remove app/jfsd/sdfljks/sdfsjl/template.hbs
```

This worked in the past because `fs.remove` is a no-op for non-existent files. Now we are calling `fs.readdir` to get folder contents, so we have to handle the exception somewhere. Instead of failing on the first `fs.readdir` deep in promises and short-circuiting, I return as soon as possible on the failure case using `fs.stat`.